example/src/MutableReference.hs

```diff
@@ -24,209 +21,173 @@
 -----------------------------------------------------------------------------
 
 module MutableReference
-  ( MemStep(..)
+  ( Action(..)
   , Problem(..)
-  , gen
+  , Ref(..)
+  , precondition
+  , transition
+  , initModel
+  , generator
   , shrink1
-  , prop_sequential
-  , prop_parallel
+  , prop_references
+  , prop_referencesParallel
   ) where
 
 import           Control.Concurrent
                    (threadDelay)
-import           Control.Monad.IO.Class
-                   (MonadIO, liftIO)
+import           Data.Functor.Classes
+                   (Eq1(..), Show1(..), showsPrec1)
 import           Data.IORef
                    (IORef, atomicModifyIORef', newIORef, readIORef,
                    writeIORef)
-import           Data.Map
-                   (Map)
-import qualified Data.Map                as M
-import           Data.Singletons.Prelude
-                   (type (@@), ConstSym1, Proxy(..), Sing(STuple0))
 import           System.Random
                    (randomRIO)
 import           Test.QuickCheck
-                   (Gen, Property, arbitrary, frequency, ioProperty,
-                   property, shrink)
+                   (Property, arbitrary, elements, frequency,
+                   ioProperty, property, shrink, (===))
 
 import           Test.StateMachine
 
 ------------------------------------------------------------------------
 
 -- Mutable references can be created, read from, written to and
--- incremented. Note that the @New@ action returns a @Reference '()@,
--- and all other actions take references, @refs \@\@ '()@, as arguments
--- -- these are special library constructs that help you work with
--- actions that have a reference-like nature.
+-- incremented.
 
-data MemStep :: Signature () where
-  New   ::                       MemStep refs ('Reference '())
-  Read  :: refs @@ '() ->        MemStep refs ('Response Int)
-  Write :: refs @@ '() -> Int -> MemStep refs ('Response   ())
-  Inc   :: refs @@ '() ->        MemStep refs ('Response   ())
+data Action (v :: * -> *) :: * -> * where
+  New   ::          Action v (Opaque (IORef Int))
+  Read  :: Ref v -> Action v Int
+  Write :: Ref v -> Int -> Action v ()
+  Inc   :: Ref v -> Action v ()
+
+deriving instance Eq1 v => Eq (Action v resp)
+
+data Ref v = Ref (v (Opaque (IORef Int)))
+
+unRef :: Ref Concrete -> IORef Int
+unRef (Ref (Concrete (Opaque ref))) = ref
+
+instance Eq1 v => Eq (Ref v) where
+  Ref v1 == Ref v2 = liftEq (==) v1 v2
+
+instance Show1 v => Show (Ref v) where
+  show (Ref v) = showsPrec1 10 v ""
+
+instance Show (Internal Action) where
```

Comment:
Yeah something like this would be nice; I'll keep it in mind as I refactor the types module (still not happy with it).